src/meta-srv/src/procedure/region_follower/create.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::{Duration, Instant};
+
+use api::v1::meta::MailboxMessage;
+use common_meta::distributed_time_constants::REGION_LEASE_SECS;
+use common_meta::instruction::{Instruction, InstructionReply, OpenRegion, SimpleReply};
+use common_meta::key::datanode_table::RegionInfo;
+use common_meta::peer::Peer;
+use common_meta::RegionIdent;
+use common_telemetry::info;
+use snafu::ResultExt;
+use store_api::storage::RegionId;
+
+use super::Context;
+use crate::error::{self, Result};
+use crate::handler::HeartbeatMailbox;
+use crate::service::mailbox::Channel;
+
+/// Uses lease time of a region as the timeout of opening a follower region.
+const OPEN_REGION_FOLLOWER_TIMEOUT: Duration = Duration::from_secs(REGION_LEASE_SECS);
+
+pub(crate) struct CreateFollower {
+    region_id: RegionId,
+    // The peer of the datanode to add region follower.
+    peer: Peer,
+}
+
+impl CreateFollower {
+    pub fn new(region_id: RegionId, peer: Peer) -> Self {
+        Self { region_id, peer }
+    }
+
+    /// Builds the open region instruction for the region follower.
+    pub(crate) async fn build_open_region_instruction(
+        &self,
+        region_info: RegionInfo,
+    ) -> Result<Instruction> {
+        let datanode_id = self.peer.id;
+        let table_id = self.region_id.table_id();
+        let region_number = self.region_id.region_number();
+
+        let RegionInfo {
+            region_storage_path,
+            region_options,
+            region_wal_options,
+            engine,
+        } = region_info;
+
+        let region_ident = RegionIdent {
+            datanode_id,
+            table_id,
+            region_number,
+            engine,
+        };
+
+        let open_instruction = Instruction::OpenRegion(OpenRegion::new(
+            region_ident,
+            &region_storage_path,
+            region_options,
+            region_wal_options,
+            true,
+        ));
+
+        Ok(open_instruction)
+    }
+
+    /// Sends the open region instruction to the datanode.
+    pub(crate) async fn send_open_region_instruction(
```

Comment:
we need to register the opening retgion to `opening_region_keeper`

Reply:
will leave a todo here